### PR TITLE
implemented the --Clear parameter.

### PR DESCRIPTION
**Pull request by dITySoftware**

### CODE REVIEW

File: source/root/usr/lib/logdata-anomaly-miner/AMiner [hunk 1]
```diff
@@ -188,6 +197,46 @@ def main():
...
+            print('The --Config and --Remove arguments must not be used together!', file=sys.stderr)
+            sys.exit(1)
+        import shutil
+        persistence_dir_name = aminer_config.config_properties.get(AMinerConfig.KEY_PERSISTENCE_DIR, AMinerConfig.DEFAULT_PERSISTENCE_DIR)
+        for filename in os.listdir(persistence_dir_name):
+            file_path = os.path.join(persistence_dir_name, filename)
+            try:
+                if not os.path.isdir(file_path):
+                    print('The AMiner persistence directory should not contain any files. Stopping..', file=sys.stderr)
+                    sys.exit(1)
```
Comment (whotwagner) [inline source/root/usr/lib/logdata-anomaly-miner/AMiner:211]: do we really have to exit here or is it possible to make it more fault tolerant?

File: source/root/usr/lib/logdata-anomaly-miner/AMiner [hunk 2]
```diff
@@ -165,6 +167,13 @@ def main():
         if arg_name == '--RunAnalysis':
             run_analysis_child_flag = True
             continue
+        if arg_name == '--Clear':
```
Comment (whotwagner) [inline source/root/usr/lib/logdata-anomaly-miner/AMiner:170]: We need another Flag "--FromBegin" that deletes just the AnalysisChild

File: source/root/usr/lib/logdata-anomaly-miner/AMiner [hunk 3]
```diff
@@ -188,6 +197,46 @@ def main():
         print("Config-Error: %s" % e)
         sys.exit(1)
 
+    if clear_persistence_flag:
+        if remove_persistence_dirs is not None:
```
Comment (whotwagner) [inline source/root/usr/lib/logdata-anomaly-miner/AMiner:201]: Please make sure that Clear does not delete the AnalysisChild

File: source/root/usr/lib/logdata-anomaly-miner/AMiner [hunk 4]
```diff
@@ -188,6 +197,46 @@ def main():
...
+                    sys.exit(1)
+                shutil.rmtree(file_path)
+            except Exception as e:
+                print('Failed to delete %s. Reason: %s' % (file_path, e), file=sys.stderr)
+
+    if remove_persistence_dirs is not None:
+        import shutil
+        persistence_dir_name = aminer_config.config_properties.get(AMinerConfig.KEY_PERSISTENCE_DIR, AMinerConfig.DEFAULT_PERSISTENCE_DIR)
+        if remove_persistence_dirs.startswith('[') and remove_persistence_dirs.endswith(']'):
+            remove_persistence_dirs = remove_persistence_dirs.rstrip(']').lstrip('[').split(',')
```
Comment (whotwagner) [inline source/root/usr/lib/logdata-anomaly-miner/AMiner:220]: We would prefer to make it more simpler without brackets "[ and ]" and please make sure that directory traversal is not possible

**Review by whotwagner — CHANGES_REQUESTED** [review]

File: source/root/usr/lib/logdata-anomaly-miner/AMiner (hunk 1 above)
Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/AMiner:211]: the aminer persistence directory should not contain files, only directories. There is a chance that the user has chosen a wrong directory. We could continue, but it is risky for the user.

File: source/root/usr/lib/logdata-anomaly-miner/AMiner (hunk 4 above)
Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/AMiner:220]: Done.

File: source/root/usr/lib/logdata-anomaly-miner/AMiner (hunk 3 above)
Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/AMiner:201]: It does not.

File: source/root/usr/lib/logdata-anomaly-miner/AMiner (hunk 2 above)
Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/AMiner:170]: I have implemented it.

**Review by whotwagner — CHANGES_REQUESTED** [review]

File: source/root/usr/lib/logdata-anomaly-miner/AMiner [hunk 5]
```diff
@@ -188,6 +201,44 @@ def main():
         print("Config-Error: %s" % e)
         sys.exit(1)
 
+    if clear_persistence_flag:
+        if remove_persistence_dirs:
+            print('The --Config and --Remove arguments must not be used together!', file=sys.stderr)
```
[the remark this answers is not in the document]
Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/AMiner:206]: yes we do. That was only a typo. --Clear was meant.

File: source/root/usr/lib/logdata-anomaly-miner/AMiner [hunk 6]
```diff
@@ -165,6 +168,16 @@ def main():
         if arg_name == '--RunAnalysis':
             run_analysis_child_flag = True
             continue
+        if arg_name == '--Clear':
+            clear_persistence_flag = True
+            continue
+        if arg_name == '--Remove':
+            remove_persistence_dirs.append(sys.argv[arg_pos])
```
[the remark this answers is not in the document]
Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/AMiner:175]: added checks in lines 175-177.

File: source/root/usr/lib/logdata-anomaly-miner/AMiner [hunk 7]
```diff
@@ -188,6 +201,44 @@ def main():
...
+            file_path = os.path.join(persistence_dir_name, filename)
+            try:
+                if not os.path.exists(file_path):
+                    print('Persistence directory %s not found!' % file_path, file=sys.stderr)
+                    continue
+                if not os.path.isdir(file_path):
+                    print('The AMiner persistence directory should not contain any files. Stopping..', file=sys.stderr)
+                    sys.exit(1)
+                shutil.rmtree(file_path)
+            except Exception as e:
```
[the remark this answers is not in the document]
Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/AMiner:233]: changed that to continue.

File: source/root/usr/lib/logdata-anomaly-miner/AMiner (hunk 1 above)
Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/AMiner:215]: Changed the AMiner to continue instead.

File: source/root/usr/lib/logdata-anomaly-miner/AMiner [hunk 8]
```diff
@@ -188,6 +204,43 @@ def main():
...
+        import shutil
+        persistence_dir_name = aminer_config.config_properties.get(AMinerConfig.KEY_PERSISTENCE_DIR, AMinerConfig.DEFAULT_PERSISTENCE_DIR)
+        for filename in os.listdir(persistence_dir_name):
+            file_path = os.path.join(persistence_dir_name, filename)
+            try:
+                if not os.path.isdir(file_path):
+                    print('The AMiner persistence directory should not contain any files.', file=sys.stderr)
+                    continue
+                shutil.rmtree(file_path)
+            except Exception as e:
```
Comment (whotwagner) [inline source/root/usr/lib/logdata-anomaly-miner/AMiner:220]: Please specify what Exception you do expect

File: source/root/usr/lib/logdata-anomaly-miner/AMiner [hunk 9]
```diff
@@ -188,6 +204,43 @@ def main():
...
+        for filename in remove_persistence_dirs:
+            file_path = os.path.join(persistence_dir_name, filename)
+            try:
+                if not os.path.exists(file_path):
+                    continue
+                if not os.path.isdir(file_path):
+                    print('The AMiner persistence directory should not contain any files.', file=sys.stderr)
+                    continue
+                shutil.rmtree(file_path)
+            except Exception as e:
```
Comment (whotwagner) [inline source/root/usr/lib/logdata-anomaly-miner/AMiner:235]: Please specify what Exception you do expect(in line 235)

**Review by whotwagner — CHANGES_REQUESTED** [review]

File: source/root/usr/lib/logdata-anomaly-miner/AMiner (hunk 8 above)
Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/AMiner:220]: shutil.rmtree actually raises an "Exception" when calling os.lstat and os.open. There is no more specific exception raised. Therefor "Exception" has to be catched here.

File: source/root/usr/lib/logdata-anomaly-miner/AMiner (hunk 9 above)
Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/AMiner:235]: same as above

File: source/root/usr/lib/logdata-anomaly-miner/AMiner (hunk 8 above)
Comment (whotwagner) [reply source/root/usr/lib/logdata-anomaly-miner/AMiner:220]: os.lstat and os.open raise OSError( https://docs.python.org/3/library/os.html ) so I believe that shutil.rmtree also raises OSError.

**Review by whotwagner — CHANGES_REQUESTED** [review]

File: source/root/usr/lib/logdata-anomaly-miner/AMiner (hunk 8 above)
Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/AMiner:220]: Changed now.

**Review by whotwagner — APPROVED** [review]

**Pull request merged**